File: python/67_Add_Binary.py
'''
Time: O(n)
'''

class Solution:
    def addBinary(self, a: str, b: str) -> str:
        res = ""
        carry = 0
        a, b = a[::-1], b[::-1]  # reverse a and b

        for i in range(max(len(a), len(b))):  # a and b may have different length
            digitA = ord(a[i]) - ord('0') if i < len(a) else 0  # convert str to int
            digitB = ord(b[i]) - ord('0') if i < len(b) else 0

            total = digitA + digitB + carry
            char = str(total % 2)
            res = char + res  # (res + char) is wrong
            carry = total // 2  # total = 2 or 3, carry = 1; total = 1, carry = 0

        if carry:  # if carry != 0
            res = '1' + res
            
        return res 



        

File: javascript/bit/67_Add_Binary.js
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
 var addBinary = function(a, b) {
    let res = "";
    let carry = 0;
    const aRev = a.split("").reverse().join(""), bRev = b.split("").reverse().join("")

    for (let i=0; i<Math.max(a.length, b.length); i++) {
        i < a.length ?  digitA = aRev[i].charCodeAt(0) - '0'.charCodeAt(0) : digitA = 0;
        i < b.length ?  digitB= bRev[i].charCodeAt(0) - '0'.charCodeAt(0) : digitB = 0;

        const total = digitA + digitB + carry;
        const char = (total % 2).toString();
        res = char + res;
        carry = Math.floor(total / 2);
    }
    if (carry) res = '1' + res;

    return res;
};

